### ISSUE TITLE
Code-first Prisma schema API

### ISSUE THREAD

**Issue by nikolasburk** [prisma/prisma#227]
As requested in [this issue](https://github.com/prisma/prisma2/issues/222), we're currently debating if/how we should provide a code-first API to build up a Prisma schema.

Citing the [original arguments](https://github.com/prisma/prisma2/issues/222#issue-471806388) (pro code-first) from [@HosseinAgha](https://github.com/HosseinAgha):

- Please re read your article on [the problems of schema first development](https://www.prisma.io/blog/the-problems-of-schema-first-graphql-development-x1mn4cb0tyl3).
- I worry that in time you reach the same conclusion and you end up with a schema language that tries to solve everything but is not as good as a real programming language.
- I think having a real general purpose programming language at your disposal is always a good idea. (take Webpack config or many other tools for example)
- I understand that Prisma is a cross language tool (Go driver for example) but it is also written in JS and Generator and Lift need node to run. So I see no harm in having a  `schema.prisma.js` option.

Note that a good foundation for a code-first API might be the [Prisma SDK](https://github.com/prisma/prisma2/issues/56). This issue is used to discuss the possibilities and pros/cons of the approaches.

**Comment by janpio** [comment]
A note: The specific problem in the issue you linked to might also be solved by adding a second param `default` (of `fallback`) to the `env()` method.

**Comment by pantharshit00** [comment]
I think if provide this approach as an option the schema file should be a generated artefact anyways to get a better overview of database structure in a readable format.

**Comment by Akkuma** [comment]
One thing pointed out in that issue is with env variables, which I've ran into playing with prisma2 is that docker postgres & prisma2 would use different env variables. This means I either need another env variable composed of my docker postgres variables, I need to combine my env variables to generate a new env variable for prisma2, or the schema needs to support multiple env variables at once so I could build the url inline.

If this were just a js file, I could load my normal .env files and build up the postgres url without resorting to needing new schema features, a .env loader that could build from other variables, or providing another env variable.

**Comment by chrisui** [comment]
Hey! Been asked to raise my use case for, at least optionally, being able to not embed env/config variables into bundles. I originally raised this in https://github.com/prisma/prisma2/issues/287#issuecomment-516388913.

In our pipeline we only ever generate _stateless builds_. This means is that you can take a build from a development branch, through staging and production without ever rebuilding. Ie. The code you develop, test and QA is exactly the same as what your users are exposed to in production.

In practice this essentially boils down to _never compile configuration into your builds_ and simply allow runtime configuration - commonly through environment variables. For example see the [recommendation on heroku](https://devcenter.heroku.com/articles/pipelines#design-considerations) for their pipleines.

Right now `prisma generate` breaks these design principles by forcing the evaluation of `env()` expressions in the schema definition. 

It would be great to have a mode whereby when generating your runtime you could configure it to be stateless so that `env("MY_ENV")` is compiled to `process.env.MY_ENV` rather than evaluated at build time.

**Comment by HosseinAgha** [comment]
I think another benefit of a JS based schema definition (code-first approach) is that we don't need to discuss issues like "how to have a modular schema?" (#92) anymore as node.js module system already handles it for us.

**Comment by matthewmueller** [comment]
You've sparked a lively debate internally about this decision ☺️

The reason we went with a declarative format rather than a code-first format is that we're describing the layout of your datasource, we're not providing implementations for this schema (unlike graphql's resolvers). In this sense, we're more like JSON schema or terraform than GraphQL. We also want to be able to migrate your database based on this schema file.

We do understand that you may want to simply generate database clients and bring your own migration engine, so for that we're planning to introduce the Prisma SDK to work with Prisma without generating a schema.

**Comment by marcus-sa** [comment]
A decorator based approach like https://typegraph.ml does with GraphQL would be nice, so you wouldn't have to use multiple classes for the same purpose :)

**Comment by HosseinAgha** [comment]
It is very good to see that you value community opinions so much.  
@matthewmueller I understand the points you make here. The Lift needs a static snapshot of current schema to make migrations.  
I still think having a JavaScript compatible schema like JSON schema is a good idea (at least as an alternative to current schema language). Lift can simply save a snapshot of the JSON object to the migrations folder.  
I think the best and most successful example for my idea is Webpack config file. Finally you can generate the prisma schema file for readability.   

IMHO by introducing a new schema definition language: 
  - the developers have to learn yet another language
  - you have to support that language and add features to it
  - only benefit here is readability and more expressive schema declaration

At the end of the day it is the trade-off that you as the platform developers should consider.

**Comment by jedmao** [comment]
Please support a TypeScript, code-first schema!

## Benefits of a TypeScript, code-first schema
1. [DRY](https://en.wikipedia.org/wiki/Don%27t_repeat_yourself), single source of truth.
1. Learn one language. Apply it everywhere. TypeScript all the things!
1. It's the missing link to writing a truly full-stack TypeScript application (see above).
1. Improve confidence by generating TS definitions for client-side applications (and more).
1. We will have achieved nirvana, the ultimate state of enlightenment.

**Comment by terion-name** [comment]
I don't like this style of definitions:
```
const User = objectType('User', t => {
  t.id('id')
  t.string('name', { nullable: true })
})

const Query = objectType('Query', t => {
  t.field('user', 'User', {
    args: { id: idArg() },
    resolve: (_, { id })
  })
})
```

It is complex, needs a lot of learning from developer, it's not really easy readable, etc.

I loved SDL-first approach paired with Prisma and graphql-import: this allowed me to make a toolset that allows to generate resolvers over schema and spin up services with rich CRUD interface in matter of minutes after data definition.

Now with lift, photon and nexus it seems I can't do this, and this is a big problem for migration of my toolsets to Prisma2

One thing that I think is mandatory is for photon to generate not only programmatic crud-client, but crud schema also, to make it possible to import crud definitions in application schema. This saves A TON of time, seriously

**Comment by vladimirdjurdjevic** [comment]
I agree with @marcus-sa. I use NestJS, which has support for GraphQL and uses TypeGraphQL and it's awesome. It allows us to have a single source of truth, which can be shared between backend and frontend. Now with Prisma2 schema, I have to duplicate everything and name my typescript classes and Prisma types differently to avoid the collision. If this could be optional (schema or code first), it would be awesome :)

**Comment by janpio** [comment]
Note: @schickling will talk to @divyenduz about this

**Comment by divyenduz** [comment]
Thanks everyone for a fruitful discussion around this topic, we are planning to have a [nexus style](https://nexus.js.org/) API for Prisma schema generation. 
 
It might look something like this (**NOTE:** This API is not final): 

```ts
import { modelType, enumType, scalarType } from 'prisma-sdk';

const Post = modelType({
  name: "Post",
  definition(t) {
    t.id("id")
    t.string("title")
    t.string("description")
    t.field("author", { type: User })
    t.field("postStatus", {
      type: POST_STATUS,
      nullable: true,
    })
  }
})
const User = objectType({
  name: "User",
  definition(t) {
    t.id("id")
    t.string("name")
    t.list.field("posts", { type: Post })
  }
})

const POST_STATUS = enumType({
  name: "POST_STATUS",
  members: ['DRAFT','PUBLISHED'],
});
```

This object structure would also be capable of "printing" an equivalent Prisma schema as well. To make other tools work with it. 

This should address the problems raised so far. What do you folks think about this as the API?

**Comment by matepapp** [comment]
it would completely make sense if the team/developer also uses graphql-nexus AND they don't have to redefine the model types again. on the other hand, it could be strange for the ones who use an SDL first approach. I'm not saying it's bad, but with this direction, you could potentially segregate your market and create an opinionated framework around the code-first approach. If this is your future plan, then great! 👍 

From my personal (and our team) perspective, we really like this syntax and it would save us a ton of time (since we're using graphql-nexus).

**Comment by divyenduz** [comment]
Thanks @matepapp 

Just to clarify, this syntax does not depend on GraphQL nexus at all, I am sorry that it wasn't clear enough. It would use a similarly looking API but nexus won't be a dependency at all. 

Essentially, there would be two ways to have a Prisma schema: 

1. Write it manually
2 Generate it via this code first API

**Comment by matepapp** [comment]
thanks for the clarification @divyenduz! 

although I thought graphql-nexus itself wouldn't be a dependency, you just wanted to point out a similar code-first solution. 
the second part of your answer helped a lot to make the thing clear. it would be great if we can decide between code-first and manual schema construction. on the other hand, you have to maintain, teach and develop 2 tools, which might be a pain in the ass in the future.

**Comment by vladimirdjurdjevic** [comment]
I prefer decorator approach like: https://typegraphql.ml/

It's much cleaner and defined classes could be reused for logic in resolvers and possibly even in frontend code. It can also generate SDL for other tooling to work. With nexus style, i don't see any advantage besides spliting definitions in multuple files. Its just another way to write SDL.

Please check out: https://docs.nestjs.com/graphql/quick-start

**Comment by HosseinAgha** [comment]
Just to make something clear hear:  
Decorator approach that everyone talking about is a very typescript specific approach. As prisma is a cross language tool I don't think depending prisma CLI on typescript in addition to node is a good idea. IMHO it is not a good idea for any node.js tool (that isn't TypeScript specific like TypeORM of TypeGraphQL) to depend on anything more than node.js and npm. Imagine Webpack or Jest to require TypeScript to work!   
There is although an exception if node.js somehow decided to add typescript support to its runtime like what [deno](https://github.com/denoland/deno) is doing.

**Comment by divyenduz** [comment]
Thanks for all the feedback, one of the early ideas of how code first Prisma schema generation might look like will live here: https://github.com/prisma/prisma-sdk-js/issues/5

There is not a lot of information there right now but that is the tracking issue.

**Comment by Weakky** [comment]
Hey everyone 👋,

We've just published a draft spec that you can find here https://github.com/prisma/prisma-sdk-js/issues/5

There are several proposals. Keep in mind none of them are definitive, we'd love to hear feedback from you first 🙌

**Comment by MichalLytek** [comment]
> As prisma is a cross language tool

So you need an SDK for every supported language?

> Imagine Webpack or Jest to require TypeScript to work!

Imagine Golang & Prisma 2 users that has to install Node.js to make Prisma Generator and Lift working 😜 

> Decorator approach that everyone is talking about is a very TypeScript specific approach.

So create a low-level SDK that others can use to create an abstraction on top of that? Just like with `graphql-js` - nobody uses it directly in apps, every framework/lib uses that under the hood.

> A decorator based approach like typegraphql.ml does with GraphQL types would be nice, so you wouldn't have to use multiple classes for the same purpose :)

@marcus Will see 😉

**Comment by HosseinAgha** [comment]
> Imagine Golang & Prisma 2 users that has to install Node.js to make Prisma Generator and Lift working 😜  
Thats exactly why I think adding TypeScript to the mix makes using Prisma 2 even harder for none `NodeJS` users. 😃  
> So create a low-level SDK that others can use to create an abstraction on top of that? Just like with graphql-js - nobody uses it directly in apps, every framework/lib uses that under the hood.  
I agree. I think the decorator based approach should be mentioned as an alternative approach and not the first recommendation in the docs.

**Comment by Weakky** [comment]
Hey @MichalLytek and @HosseinAgha 👋, and thanks for providing feedback!

> So create a low-level SDK that others can use to create an abstraction on top of that? Just like with graphql-js - nobody uses it directly in apps, every framework/lib uses that under the hood.

I'm not sure I'd qualify `graphql-js` as an easy-to-use low-level abstraction to build tools on top of it. Having these nested/wrapped classes to describe types, nullability & lists is IMO incredibly painful to work with.

> Maybe SDK should have a simple, object-config API and then have nexus/decorators wrappers around that (some maintained by community)?

> I agree. I think the decorator based approach should be mentioned as an alternative approach and not the first recommendation in the docs

Agreed! The SDK should indeed be a solid foundation for tools to be built on top of it, but we also think it should provide _a decent DX_ for users who want to use it as-is (preferring the code-first approach over the Prisma Schema Language)

Therefore, would you mind elaborating on why the following proposal https://github.com/prisma/specs/tree/spec/prisma-schema-sdk/sdk-js/schema#1-nexus-style-attributes-as-objects is not suited as a low-level abstraction? 

While we could indeed go with an even simpler API, we're interested in hearing use-cases that you could not accomplish with the proposal linked just above!

Thanks 🙌

**Comment by MichalLytek** [comment]
> The point is, if you ever need a "very" low-level API, you could still construct the AST manually, in any language.

And that's great news! AST and producing config objects is always simpler than using an API that is designed for DX (writing by hand) rather than generating from config (metadata) by calling some callbacks or decorators.

So basically the SDK would expose a proper interfaces for datamodel and a `render` method that produces a formatted PSL string that we can write to `schema.prisma` file as a code-first approach?

**Comment by jasonkuhrt** [comment]
Hey @weaky, nice work so far!

Have you explored the idea of combinators? It has some similarities to chaining but more general and higher degrees of composition.

Here's a sketch of what that might look like. It takes up the `Post` data in your fluent example.

```ts
pipeValue(
  model('Post'),
  pipe(int('id'), primary),
  string('title'),
  boolean('draft'),
  pipe(string('categories'), list),
  string('slug'),
  pipe(field('author'), ref(() => User)),
  pipe(field('comments'), ref(() => Comment), list),
  pipe(dateTime('createdAt'), default(now)),
  pipe(dateTime('updatedAt'), updatedAt(true)),
)
```

Via composition the flexibility of static combinators are very high and the semantics are clear. Repeated patterns can be factored out naturally.

```ts
const createdAt = pipe(dateTime('createdAt'), default(now))

pipe(
  model('Post'),
  createdAt,
  ...
)

pipe(
  model('User'),
  createdAt
)
```

Combinators are like lego.

The `pipe`s could be reimagined using the [upcoming pipe operator(s)](https://babeljs.io/blog/2019/07/03/7.5.0).

```ts
model('Post') |> (
  (int('id') |> primary)
  string('title'),
  boolean('draft'),
  (string('categories') |> list),
  string('slug'),
  (field('author') |> ref(() => User)),
  (field('comments') |> ref(() => Comment), list),
  (dateTime('createdAt') |> default(now)),
  (dateTime('updatedAt') |> updatedAt(true)),
)
```

But I'm not seeing such a clear improvement here, personally. Anyways, something to note.

This is primarially an example, to convey the gist. I do not have an explanation of how it could work under the hood presently. For that I would personally need to conduct some prototypes.

This API could be layered with a chaining API if so desired, without too much effort I think. Its a matter of creating the fluency (always return `this`) and each static having its state param pre-filled with `this`.

A key point here is that the combinators would be working on an underlying immutable AST, building it up incrementally. Put simply, the combinators are just sugar saving the developer from having to write POJOs themselves.

**Comment by jasonkuhrt** [comment]
> A key point here is that the combinators would be working on an underlying immutable AST, building it up incrementally. Put simply, the combinators are just sugar saving the developer from having to write POJOs themselves.

@Weakky et al. clarified offline that this is the case. There is/would be a raw AST represented as POJOs in JS. Given a spec for this, then any user could create any tool/library whatever to produce it.

While the Prisma SDK will ship with its vision of a tool, the AST invites other approaches.

**Comment by Weakky** [comment]
The spec was updated. We added

- A diagram to better explain the architecture
- A link towards a separate issue that shows an early version of the AST generated by the Prisma Schema SDK (and possibly third-party Prisma Schema SDKs)
- A clear mention that explains what the proposals are for

**Comment by tvvignesh** [comment]
Hi guys. Went through the entire thread in this discussion and [this](https://www.prisma.io/blog/the-problems-of-schema-first-graphql-development-x1mn4cb0tyl3) article as well. Had some things to raise from my end.

SDL first is very readable, easy to define and can be understood by anyone with just basic programming knowledge and also the fact that its not good to tie Prisma to a language in my humble opinion. For instance, if you see the example [here](https://nexus.js.org/converter) the SDL looks neat and is just 70 lines of code while code-first approach is not so readable and takes 140 lines of code.

Also, the problem with modularizing schema can be addressed in ways like what Apollo did with Federation enabling developers to still use SDL but allow modularising code.

If I am right (not sure), this would bring an additional build step to convert code to SDL - is it really needed?

The main advantage I see personally in code-first is the ability to use the language-specific powerful features to define types with no need to define your own schema language.

My opinion is that, the option should be left to the developer (either choose SDL first or code first) rather than the library taking the call thus giving some flexibility. Since the SDL is anyways going to be generated as an artifact from the code-first API, its better to have both.

The way I see it, the problem is not necessarily in the SDL but in the ability of the IDE extensions to parse through the schema and provide autocompletion, type suggestions, etc. I am not sure, but isn't it better to add such capabilities to the extension rather than going for a different approach?

May be we have to consider this like what GraphQL did - Prisma the specification and Prisma the tooling around it - that way, there will be a consistent schema language for all tools which want to implement the Prisma spec (What if new ORMs come in future, which are not written in Node.js but wants to use the Prisma specification alone - it becomes tricky in such cases) and the Prisma tooling which is currently written in Node.js

Thanks.

**Comment by Aperrix** [comment]
Still no news regarding the development of this feature or an alternative?
I love everything that has been done on prisma2, i would really like to migrate 😞

**Comment by Ericnr** [comment]
This would be my preferred way to define the schema too

**Comment by vjsingh** [comment]
Agreed. After Prisma pushed for a code-first approach with Nexus this summer, I moved our project over to it. I like code-first and this is a blocker for me as well in terms of migrating to Prisma2.

**Comment by jedmao** [comment]
They [release a preview roughly every 2 weeks](https://github.com/prisma/prisma2/releases) if you want to get your hands dirty.

**Comment by iddan** [comment]
In the midtime, we released a library implementing the proposed AST and code generator: https://github.com/amplication/prisma-schema-dsl

**Comment by janpio** [comment]
Duplicate of this issue here: https://github.com/prisma/prisma/issues/1396#issuecomment-666764195

**Comment by dustinlacewell** [comment]
Can there be a status update? I have a project which would really benefit from a end-to-end code-first approach.

**Comment by omerp-explorium** [comment]
😢

**Comment by alexn-s** [comment]
is there any progress on this manner?

**Comment by alexn-s** [comment]
for anybody still looking for a solution:

 i found a library https://github.com/ridafkih/schemix that
> allows you to programmatically create Prisma schemas using TypeScript

**Comment by igalklebanov** [comment]
Hey 👋

Cloaing this for staleness and being dusty.
